transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollChannel.java

```diff
@@ -383,6 +384,43 @@ protected final int doWriteBytes(ChannelOutboundBuffer in, ByteBuf buf) throws E
         return WRITE_STATUS_SNDBUF_FULL;
     }
 
+    /**
+     * Write bytes to the socket, with or without a remote address.
+     * Used for datagram and TCP client fast open writes.
+     */
+    protected final long doWriteOrSendBytes(ByteBuf data, InetSocketAddress remoteAddress, boolean fastOpen)
+            throws IOException {
+        if (data.hasMemoryAddress()) {
+            long memoryAddress = data.memoryAddress();
+            if (remoteAddress == null) {
+                return socket.writeAddress(memoryAddress, data.readerIndex(), data.writerIndex());
```

Comment:
Would it be correct to put `assert !fastOpen` here?

Reply:
Sounds right. hmm, not sure if we should do an assert or an IllegalArgumentException check?  Either one can probably go at that top to just check that remoteAddress must be set if fastOpen is true.

Reply:
also, just noticed that this method i moved up (normally used for datagrams) is pretty much the same as the normal method used for TCP (doWriteBytes), with the exception of writev. Does writev not work with TCP? if it should work, those methods can probably be combined

Reply:
hmmm, yea Socket.writev and Socket.writevAddresses appear to do mostly the same thing. will resist trying to condense those for now

Reply:
yea, they seem to be the same, writev just makes a native iovec, writevAddress relies on an externally built one. i can condense those in this PR if it makes sense